### ISSUE TITLE
Resolve nested type references in proto compiler

### ISSUE THREAD

**Issue by mourner** [mapbox/pbf#25]
Currently the proto compiler doesn't properly resolve types like `Feature.Type`. Also, same-name types in different namespaces lead to broken code.

Resolving types properly will add some complexity in the compiler code, but it's necessary.

**Comment by mikemorris** [comment]
Hit this while working on a new proto spec for https://github.com/mapbox/node-fontnik, worked around the nesting/same-types issue by simply using a `FaceMetrics` message type instead of a nested `Metrics` message inside the `Face`.

**Broken**
```
message Face {
    optional Metrics metrics = 1;

    optional Metrics {
        optional double ascender = 1;
        optional double descender = 2;
    }
}
```

**Workaround**
```
message Face {
    optional FaceMetrics metrics = 1;
}

optional FaceMetrics {
    optional double ascender = 1;
    optional double descender = 2;
}
```

FWIW, naming the non-nested class `Face_Metrics` generates roughly equivalent C++ headers when running the proto spec through `protoc`.

**Comment by mourner** [comment]
@RReverser has a great suggestion: to use `Object.create` for tracking message scopes, potentially making it easy to implement proper nested type resolution.

**Comment by mourner** [comment]
Finally implemented this, pushing soon! Along with a fix for #34.
